components/style/values/specified/box.rs

```diff
@@ -492,3 +493,82 @@ pub fn assert_touch_action_matches() {
         NS_STYLE_TOUCH_ACTION_MANIPULATION => TouchAction::TOUCH_ACTION_MANIPULATION,
     }
 }
+
+bitflags! {
+    #[derive(MallocSizeOf, ToComputedValue)]
+    /// Constants for contain: https://drafts.csswg.org/css-contain/#contain-property
+    pub struct Contain: u8 {
+        /// `layout` variant, turns on layout containment
+        const LAYOUT = 0x01;
+        /// `style` variant, turns on style containment
+        const STYLE = 0x02;
+        /// `paint` variant, turns on paint containment
+        const PAINT = 0x04;
+        /// `strict` variant, turns on all types of containment
+        const STRICT = 0x8;
+        /// variant with all the bits that contain: strict turns on
+        const STRICT_BITS = Contain::LAYOUT.bits | Contain::STYLE.bits | Contain::PAINT.bits;
+    }
+}
+
+impl ToCss for Contain {
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+        if self.is_empty() {
+            return dest.write_str("none")
+        }
+        if self.contains(Contain::STRICT) {
+            return dest.write_str("strict")
+        }
+
+        let mut has_any = false;
+        macro_rules! maybe_write_value {
+            ($ident:path => $str:expr) => {
+                if self.contains($ident) {
+                    if has_any {
+                        dest.write_str(" ")?;
+                    }
+                    has_any = true;
+                    dest.write_str($str)?;
+                }
+            }
+        }
+        maybe_write_value!(Contain::LAYOUT => "layout");
+        maybe_write_value!(Contain::STYLE => "style");
+        maybe_write_value!(Contain::PAINT => "paint");
+
+        debug_assert!(has_any);
+        Ok(())
+    }
+}
+
+impl Parse for Contain {
+    /// none | strict | content | [ size || layout || style || paint ]
+    fn parse<'i, 't>(
+        _context: &ParserContext,
+        input: &mut Parser<'i, 't>
+    ) -> Result<Contain, ParseError<'i>> {
+        let mut result = Contain::empty();
+        while let Ok(name) = input.try(|i| i.expect_ident_cloned()) {
+            let flag = match_ignore_ascii_case! { &name,
+                "layout" => Some(Contain::LAYOUT),
+                "style" => Some(Contain::STYLE),
+                "paint" => Some(Contain::PAINT),
+                "strict" if result.is_empty() => return Ok(Contain::STRICT | Contain::STRICT_BITS),
+                "none" if result.is_empty() => return Ok(result),
```

Comment:
I think you'll need to check `if result.is_empty()` inside the pattern instead of using it as a pattern matching [`guard`](https://rustbyexample.com/flow_control/match/guard.html). If putting `if result.is_empty()` as `guard` here, I think it will only get into the `strict` one?

IMO, just like @emilio commented in the [comment](https://github.com/servo/servo/pull/19682/files/8472565f7a7f9ba877adcbd89e8c177c33eb777a..63e7b3f5cbaf2c5652dfbecb7fee901806dfb72b#r159568931), it might be fine to be like following

```rust
"strict" => {
    if result.is_empty() {
        return Ok(Contain::STRICT | Contain::STRICT_BITS)
     }
     None
},
"none" => {
    if result.is_empty() {
        return Ok(result);
    }
    None
}
```

Reply:
Thanks, I changed it back to the previous version. I don't really understand why it won't get into the `none` arm though; how would rust evaluate the guard?

Reply:
FWIW the guard should work just fine, and it's more elegant indeed.

Wanna open a PR after this one merges simplifying it? Thanks!

Reply:
Sure, I'd be glad to, assuming this one passes the tests.

Reply:
Ahhh... yes... sorry, I think it in a wrong way so using guard here should work fine.
Sorry for the misleading 😣